runtime/src/accounts_db.rs

```diff
@@ -6511,6 +6511,56 @@ impl AccountsDb {
         self.flush_slot_cache_with_clean(&[slot], None::<&mut fn(&_, &_) -> bool>, None)
     }
 
+    /// 1.13 and some 1.14 could produce legal snapshots with more than 1 append vec per slot.
+    /// This is now illegal at runtime in the validator.
+    /// However, there is a clear path to be able to support this.
+    /// So, combine all accounts from 'slot_stores' into a new storage and return it.
+    /// This runs prior to the storages being put in AccountsDb.storage
+    pub(crate) fn combine_multiple_slots_into_one_at_startup(
+        path: &Path,
+        id: AppendVecId,
+        slot: Slot,
+        slot_stores: &HashMap<AppendVecId, Arc<AccountStorageEntry>>,
+    ) -> Arc<AccountStorageEntry> {
+        let size = slot_stores
+            .iter()
+            .map(|(_id, store)| store.accounts.capacity())
+            .sum();
+        let storage = AccountStorageEntry::new(path, slot, id, size);
+
+        // get unique accounts, most recent version by write_version
+        let mut accum = HashMap::<Pubkey, StoredAccountMeta<'_>>::default();
+        slot_stores.iter().for_each(|(_id, store)| {
+            store.accounts.account_iter().for_each(|loaded_account| {
+                let loaded_write_version = loaded_account.write_version();
+                match accum.entry(*loaded_account.pubkey()) {
+                    std::collections::hash_map::Entry::Occupied(mut occupied_entry) => {
+                        if loaded_write_version > occupied_entry.get().write_version() {
+                            occupied_entry.insert(loaded_account);
+                        }
+                    }
+
+                    std::collections::hash_map::Entry::Vacant(vacant_entry) => {
+                        vacant_entry.insert(loaded_account);
+                    }
```

Comment:
*style/subjective*

`hash_map::Entry` has a few helper functions, that could make this a bit shorter:
```suggestion
                accum.entry(*loaded_account.pubkey())
                    .and_update(|entry| {
                        let new_version = loaded_account.write_version();
                        if new_version > entry.get().write_version() {
                            *entry = loaded_account;
                        }
                    }
                    .or_insert(loaded_account);
```

Reply:
didn't work due to borrow checker on loaded_account. Cleaned the current impl up a little. I think fwiw `and_modify` is correct instead of `and_update`. Not sure why.

Reply:
Sorry, I did not check.  Wrote that code directly on GitHub.

You are right, the borrow checker does not understand that only one of those closures will be called.
While with an explicit `if`, the compiler treats each branch as exclusive, meaning that it can consume the `loaded_account` value.

There is a way to get this code to work.  If you add `#[derive(..., Copy, Clone, ...)` to the `StoredAccountMeta<'a>`:
```rust
#[derive(PartialEq, Eq, Clone, Copy, Debug)]
pub struct StoredAccountMeta<'a> {
```

And then this compiles:
```rust
        for store in slot_stores.values() {
            for loaded_account in store.accounts.account_iter() {
                accum.entry(*loaded_account.pubkey())
                    .and_modify(|entry| {
                        if loaded_account.write_version() > entry.write_version() {
                            *entry = loaded_account;
                        }
                    })
                    .or_insert(loaded_account);
            }
        };
```

Semantically, it creates two copies.  But, I imagine, after inlining the closures (which the compiler will almost certainly do) the copy will be optimized away.

I am in no way insisting that this is the way it should be done.
Just wanted to figure out why was it not working :)